sdk/src/azure/core/az_json_token.c

```diff
@@ -325,6 +325,52 @@ AZ_NODISCARD static az_result _az_json_token_get_string_helper(
   return AZ_OK;
 }
 
+/**
+ * @brief Gets the JSON token's string after unescaping it, if required.
+ */
+AZ_NODISCARD az_result az_json_string_unescape(
+    az_span json_string,
+    char* destination,
+    int32_t destination_max_size,
+    int32_t* out_string_length)
+{
+  _az_PRECONDITION_VALID_SPAN(json_string, 1, false);
+  _az_PRECONDITION_NOT_NULL(destination);
+  _az_PRECONDITION(destination_max_size > 0);
+
+  int32_t position = 0;
+  for (int32_t i = 0; i < json_string._internal.size; i++)
+  {
+    uint8_t current_char;
+    current_char = json_string._internal.ptr[i];
+    if (current_char == '\\' && i < json_string._internal.size)
+    {
+      uint8_t next_char = json_string._internal.ptr[i + 1];
+      // check that we have sometig to escape
+      if (_az_is_valid_escaped_character(next_char))
+      {
+        current_char = _az_json_unescape_single_byte(next_char);
+        i++;
+      }
```

Comment:
If it isn't a valid escaped character, please return an error `az_result`, and document that (in the else case).

We shouldn't silently ignore invalid JSON as input.

Reply:
i can't do that how do i know that "\\9" is not what the customer intended in that string. that check only sais that the next byte is a known escape character, but i cannot determine the intent of a char that is not escapable. See the definition of the escape next byte, it cannot tell if it is an erro or not. it returns the char if it is not a known one.

Reply:
I understand. That said, the API we are adding is about JSON unescaping.

> i can't do that how do i know that "\9" is not what the customer intended in that string

You know because such a string would be invalid JSON and is meaningless (or at least undefined) to try to unescape such a string. It is invalid data (or possibly a data corruption or user-side bug).

The only valid characters that can follow a `\` in a JSON string are one of this set:
https://datatracker.ietf.org/doc/html/rfc8259#section-7

We have to decide on one of the following (and document that behavior). Either:
- The user has to provide valid JSON, as an az_span
  - What happens if they don't?
- The user can pass any arbitrary input, and we validate/return an error when the input is malformed JSON.

> See the definition of the escape next byte, it cannot tell if it is an erro or not. it returns the char if it is not a known one.

That is just a helper method used to unescape already validated data. A `az_json_token` is known to be valid already, because of how it is created.

Reply:
Btw, I am talking about having an else for this: `if(_az_is_valid_escaped_character(next_char))`

That check should end up catching and correctly rejecting `\9` as invalid.